phoenix-core-server/src/main/java/org/apache/phoenix/replication/log/LogFile.java

```diff
@@ -0,0 +1,236 @@
+import java.util.Map;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hbase.Cell;
+import org.apache.hadoop.hbase.io.compress.Compression;
+import org.apache.hadoop.hbase.util.Bytes;
+
+/**
+ * Defines the structure and constants for Phoenix Replication Log files.
+ * Provides interfaces for reading and writing these logs.
+ */
+public final class LogFile {
+
+    /** Magic number for Phoenix Replication Log files */
+    public static byte[] MAGIC = Bytes.toBytes("PLOG");
+
+    /** Current major version of the replication log format */
+    public static byte VERSION_MAJOR = 1;
+    /** Current minor version of the replication log format */
+    public static byte VERSION_MINOR = 0;
+
+    /** Size of the block checksum trailer (CRC64) */
+    public static int CHECKSUM_SIZE = Bytes.SIZEOF_LONG;
+
+    /** Represents the file header */
+    public interface Header {
+        int getMajorVersion();
+        Header setMajorVersion(int majorVersion);
+        int getMinorVersion();
+        Header setMinorVersion(int minorVersion);
+        int getSerializedLength();
+        void readFields(DataInput in) throws IOException;
+        void write(DataOutput out) throws IOException;
+    }
+
+    /** Represents the header of a single block within the log file */
+    public interface BlockHeader {
+        /** Magic number for Phoenix Replication Log blocks */
+        public static final byte[] MAGIC = Bytes.toBytes("PBLK");
+        /** Current version of the replication log block header */
+        public static final byte VERSION = 1;
+        int getVersion();
+        Compression.Algorithm getCompression();
+        BlockHeader setCompression(Compression.Algorithm compression);
+        int getUncompressedSize();
+        BlockHeader setUncompressedSize(int uncompressedSize);
+        int getCompressedSize();
+        BlockHeader setCompressedSize(int compressedSize);
+        int getSerializedLength();
+        void readFields(DataInput in) throws IOException;
+        void write(DataOutput out) throws IOException;
+    }
+
+    /** Represents the file trailer */
+    public interface Trailer {
+        int getMajorVersion();
+        Trailer setMajorVersion(int majorVersion);
+        int getMinorVersion();
+        Trailer setMinorVersion(int minorVersion);
+        long getRecordCount();
+        Trailer setRecordCount(long recordCount);
+        long getBlockCount();
+        Trailer setBlockCount(long blockCount);
+        long getBlocksStartOffset();
+        Trailer setBlocksStartOffset(long offset);
+        long getTrailerStartOffset();
+        Trailer setTrailerStartOffset(long offset);
+        int getSerializedLength();
+        void readFields(DataInput in) throws IOException;
+        void write(DataOutput out) throws IOException;
+    }
+
+    /** Represents a single logical change */
+    public interface Record {
```

Comment:
A record maps to an HBase mutation. I think it will be more efficient and simpler if the Phoenix coproc through ReplicationLogWriter passes an HBase mutation directly to create a record and also gets back a mutation when ReplicationLogReader reads a record. Then the in-memory representation of a record is composed of mainly an HBase mutation, commit id, and schema and there will be no need to translate a mutation to a record outside the log file implementation.

Reply:
I agree with you on what you wrote. I did not mean to use the HBase code to serialize and deserialize mutations. Mutation is a public API object and our code should stay at that level. I observed that if the mutation is also the API for the log file library this PR builds then it would simplify the Phoenix code using this API and also simplify the library a bit, and at the same time it would make the solution more efficient. For example, the library would not need to redefine mutation type and would not need to keep track of columns using a separate map structure as mutation already has this map structure. This should reduce memory allocations.

Reply:
Ok, I am working on a commit for this PR that addresses this review feedback:
 - Document LogFile interfaces with javadoc
 - Directly accept HBase Mutation objects in the Writer
 - Directly provide HBase Mutation objects from the Reader

Additional changes:
 - Optimize the encoding and decoding of column family and qualifiers, to avoid string concat/splitting, changing the file format. Spec requires update!
 - Encode records with a single pass. Saves on lots of duplicate structure enumeration, requiring a small copy in trade.
 - LogFileCompressionTest test fix.
 
I will push it up soon.

Reply:
I pushed up the changes that resulted from this feedback as d34d4dc90 . 
I need to add more tests for serde of the various ways Put and Delete mutations might be constructed.
@kadirozde